rbs/CommentsAssociator.cc

```diff
@@ -0,0 +1,282 @@
+            if (absl::StartsWith(it->second.string, prefix)) {
+                if (auto e = ctx.beginError(it->second.loc, core::errors::Rewriter::RBSUnusedComment)) {
+                    e.setHeader("Unused RBS signature comment. No method definition found after it");
+                }
+            }
+            it = commentByLine.erase(it);
+        } else {
+            break;
+        }
+    }
+}
+
+void CommentsAssociator::consumeDanglingComments(parser::Node *node) {
+    auto nodeEndLine = core::Loc::pos2Detail(ctx.file.data(ctx), node->loc.endPos()).line;
+
+    for (auto it = commentByLine.begin(); it != commentByLine.end();) {
+        if (it->first < nodeEndLine) {
+            if (absl::StartsWith(it->second.string, RBS_PREFIX)) {
+                if (auto e = ctx.beginError(it->second.loc, core::errors::Rewriter::RBSUnusedComment)) {
+                    e.setHeader("Unused RBS signature comment. No method definition found after it");
+                }
+            }
+            it = commentByLine.erase(it);
+        } else {
+            break;
+        }
+    }
+}
+
+void CommentsAssociator::associateCommentsToLines() {
+    for (auto &[start, end] : commentLocations) {
+        auto comment_string = ctx.file.data(ctx).source().substr(start, end - start);
+        auto start32 = static_cast<uint32_t>(start);
+        auto end32 = static_cast<uint32_t>(end);
+        auto comment = CommentNode{core::LocOffsets{start32, end32}, comment_string};
+
+        auto line = core::Loc::pos2Detail(ctx.file.data(ctx), start32).line;
+        commentByLine[line] = comment;
+    }
+}
+
+void CommentsAssociator::associateCommentsToNode(parser::Node *node,
+                                                 const InlinedVector<std::string_view, 2> &prefixes) {
+    auto nodeStartLine = core::Loc::pos2Detail(ctx.file.data(ctx), node->loc.beginPos()).line;
+
+    vector<CommentNode> comments;
+
+    // fmt::print("associating comments to {}\n", node->toString(ctx));
+    for (auto it = commentByLine.begin(); it != commentByLine.end();) {
+        // fmt::print("it->first: {}, nodeStartLine: {}\n", it->first, nodeStartLine);
+        if (it->first < nodeStartLine) {
+            bool found = false;
+            for (const auto &prefix : prefixes) {
+                if (absl::StartsWith(it->second.string, prefix)) {
+                    // fmt::print("associated {} to {}\n", it->second.string, node->toString(ctx));
+                    comments.emplace_back(it->second);
+                    it = commentByLine.erase(it);
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                it++;
+            }
+        } else {
+            break;
+        }
+    }
+
+    commentsByNode[node] = comments;
+}
+
+void CommentsAssociator::walkNodes(parser::Node *node) {
+    if (node == nullptr) {
+        return;
+    }
+
+    typecase(
+        node,
+
+        // Blocks
```

Comment:
```suggestion
        // Blocks

```

Reply:
I've had a hard time categorizing these nodes so I split them between "cleans up preceding" and "doesn't clean up preceding". I think it's more useful.